### ISSUE TITLE
UDP server crashes on accessing the port

### ISSUE THREAD

**Issue by gchiu** [metaeducation/ren-c#767]
In this demo code https://gist.github.com/gchiu/4dce1551945cb1c654679bbf1eebb429

this line [read listen](https://gist.github.com/gchiu/4dce1551945cb1c654679bbf1eebb429#file-udp-server-reb-L44) is needed to start the server, and is not needed for TCP.

this line [client: first evt/port](https://gist.github.com/gchiu/4dce1551945cb1c654679bbf1eebb429#file-udp-server-reb-L44) which is used in a tcp server to get the port then crashes the UDP server

> PANIC FUNCTION! called (explicitly or by rebPanic() API)
C Source File --anonymous--, Line 0
At evaluator tick: 214887
Rebol Internal Error

**Comment by hostilefork** [comment]
Crash reproduces for me on both Windows and Linux.

There is a line attempting an [OS_FREE()](https://github.com/metaeducation/ren-c/blob/d92806094101ba61f6b4d9734ff02694cf972d7b/src/core/p-net.c#L109) on a "nsock" variable, which it gets out of [req->common.sock](https://github.com/metaeducation/ren-c/blob/d92806094101ba61f6b4d9734ff02694cf972d7b/src/core/p-net.c#L91).

So it expects that `req->common.sock` to have been OS_ALLOC()'d by someone.  It remarks "allocated by dev_net.c", but going back to the R3-Alpha repository *(where memory allocation was called OS_Make or OS_MAKE)* I don't see anything related.  The only case of an allocation that would be OS_FREE'd is in [asynchronous DNS code](https://github.com/rebol/rebol/blob/25033f897b2bd466068d7663563cd3ff64740b94/src/os/dev-net.c#L302), which is totally unrelated.

The only thing in R3-Alpha related to OS_Make that looks like it would be related would be the OS_Make in [OS_Make_Devreq](https://github.com/rebol/rebol/blob/25033f897b2bd466068d7663563cd3ff64740b94/src/os/host-device.c#L329).  But I don't see how that is making it into the common.sock field, or where it's assigned.

The call to this function is coming from [an attempt to PICK from a PORT!](https://github.com/metaeducation/ren-c/blob/master/src/core/p-net.c#L379) (FIRST being PICK 1).  It doesn't check if sock is NULL, it checks if data is NULL, and it's not.  But it's a union, so you're not supposed to read from members that aren't the one you're looking at anyway.

@zsx any ideas?  It would be nice if Graham could do his drone-control-scripting in Rebol, so just need to get the I/O part set well enough to start scripting on top of it.

**Comment by zsx** [comment]
First of all, it's not supposed to pick from a UDP port, because 'pick port 1' is used to get the new connection port, but in the case of UDP, there is not such a port. But it shouldn't crash of course, I'll push a quick fix for that.

Second, the UDP port is not well tested, and I added this because we felt that we might need it, and it turned out that we haven't had a use case for it yet, and because it's piggybacked on the TCP implementation, the user interfaces aren't well thought through, e.g. the remote IP/port-id is not exposed to the user space, which means it can't write back to the remote port directly. Even the remote IP/port-id is exposed, WRITE doesn't fit quite well for what a user needs, because in a lot of cases, a UDP server would like to write to the peer it receives the packet from, which means each write could have a different IP and port-id, and WRITE in current form only accepts data. We could create a new port for each packet it receives, but the underlying socket has be to shared by all of these ports (not sure how to do that yet), or WRITE will come from different local port-ids, which might not be what the other peer expected.

**Comment by gchiu** [comment]
In my particular scenario I am sending commands on one port, and the device is responding on a different port. So my server port just needs to accept those latter responses, but doesn't need to reply.

**Comment by hostilefork** [comment]
@zsx Graham's situation may be simple enough to be covered by a slight change in what's there.

He does `drone: open udp://192.168.10.1:8889` and then `write drone to binary! "command"`

The UDP connection apparently does a bind() and gets an OS-allocated local port, at a large number e.g. 53295.  The drone picks up on this, and writes back an "OK" response there (via UDP)

But if he says `wait drone` then it doesn't get the response.  WAIT returns immediately, with a blank, because RRF_PENDING is not set on the port.

I gather it's just not waiting long enough?  Because if he throws in a `wait 1` before the `wait drone` it's long enough to get the response to kick off an event loop.

I'm unsure what this RRF_PENDING is supposed to mean in the scheme of things, but it seems to me that if WAIT on a UDP port would sit around waiting for some kind of signal of "doneness" instead of just returning, it might solve the problem of wanting to do a UDP loop for this purpose.

**Comment by hostilefork** [comment]
I tinkered with this enough to hopefully get it going:

https://github.com/metaeducation/ren-c/commit/8f0878b8cacf1b1a8047c8c95c6bd9a6eab2e1d2

Review very welcome--since I really was just poking at it to see if I could make it work without changing too much--and I stopped poking at it once it worked for the one test I was doing.  :-/

Observing as a user, I definitely feel like you shouldn't have to do that `read` in order to get the process going; it seems like if a port is "listener class" then it should automatically be listening... (?)

> We could create a new port for each packet it receives, but the underlying socket has be to shared by all of these ports (not sure how to do that yet), or WRITE will come from different local port-ids, which might not be what the other peer expected.

It sounds like you're discussing a "peer" e.g. P2P style connection through UDP.  So kind of simulating TCP at that point (?)  I dunno.  Graham's situation is more a client server thing, so I think this should cover it.
